### PR TITLE
Upgrade Node.js to v20 and fix loading swc-node

### LINKED ISSUE #4458: Weblate tests fails in production after `xml-editor` merge
```
node:internal/process/esm_loader:40
      internalBinding('errors').triggerUncaughtException(
                                ^
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/home/runner/work/specify7/specify7/specifyweb/frontend/js_src/lib/components/DataModel/resourceApi' imported from /home/runner/work/specify7/specify7/specifyweb/frontend/js_src/lib/components/DataModel/saveBlockers.tsx
```

[saveBlockers.tsx](https://github.com/specify/specify7/blob/production/specifyweb/frontend/js_src/lib/components/DataModel/saveBlockers.tsx) in `production`, [saveBlockers.ts](https://github.com/specify/specify7/blob/v7.9.3-copy/specifyweb/frontend/js_src/lib/components/DataModel/saveBlockers.ts) in `v7.9.3-copy`

image

We'll need to pick up on this and resolve it soon

### PR DESCRIPTION
**Pull request by acwhite211**
Fixes #4458 

This PR temporarily migrates away from swc until the issues https://github.com/swc-project/swc-node/pull/748 and https://github.com/swc-project/swc-node/issues/743 are in the next release, presumably version 1.8.1.  This fix will be considered as a solution when the Weblate error doesn't occur in our project build/testing.

### CODE REVIEW

**Comment by ehaynes99** [comment]
Not associated with this project, just saw the link in https://github.com/swc-project/swc-node/pull/748

This was already fixed, so you should be able to replace the `--loader like:
```diff
@@ -16,16 +16,16 @@
   },
   "scripts": {
     "//": "All of these should be executed from the js_src directory",
-    "localizationTests": "node --loader @swc-node/register/esm lib/localization/__tests__/scanUsages.ts",
-    "pullFromWeblate": "node --loader @swc-node/register/esm lib/localization/utils/pull.ts",
-    "schema:extract": "node --loader @swc-node/register/esm lib/localization/schema-localization/extract.ts",
-    "schema:pullFromWeblate": "node --loader @swc-node/register/esm lib/localization/schema-localization/pull.ts",
+    "localizationTests": "node --import @swc-node/register/esm-register lib/localization/__tests__/scanUsages.ts",
+    "pullFromWeblate": "node --import @swc-node/register/esm-register lib/localization/utils/pull.ts",
+    "schema:extract": "node --import @swc-node/register/esm-register lib/localization/schema-localization/extract.ts",
+    "schema:pullFromWeblate": "node --import @swc-node/register/esm-register lib/localization/schema-localization/pull.ts",
     "test": "npm run typecheck && npm run unitTests && npm run localizationTests",
     "typecheck": "tsc",
     "unitTests": "jest",
     "unitTests:coverage": "jest --coverage",
     "unitTests:watch": "jest --watch",
-    "validateWeblate": "node --loader @swc-node/register/esm lib/localization/__tests__/validateWeblate.ts"
+    "validateWeblate": "node --import @swc-node/register/esm-register lib/localization/__tests__/validateWeblate.ts"
   },
   "dependencies": {
     "@codemirror/lang-json": "^6.0.0",
```

**Comment by maxpatiiuk** [comment]
@ehaynes99 Thanks for the comment!
@acwhite211 can you try out that solution?

we moved from ts-node to swc because ts-node started failing on new node.js version. Now swc is failing on new node.js version 🤪

**Comment by acwhite211** [comment]
Looks like because we are currently using node version 18.19.0 that we should be using `--loader` instead of `--import` as it says in the [README](https://github.com/swc-project/swc-node/blob/28c0a9969942499c2dc9a2883360146de14ef907/README.md?plain=1#L21).

**Comment by maxpatiiuk** [comment]
Can we update the CI and Dockerfile to 20 then?
According to https://nodejs.org/en/about/previous-releases, Node.js 18 is currently in maintenance mode

**Comment by acwhite211** [comment]
After updating to node v20.6 and changing `--loader @swc-node/register/esm` to `--import @swc-node/register/esm-register` in the package.json, there still is an error I'm working on resolving
```
> node --import @swc-node/register/esm-register lib/localization/__tests__/scanUsages.ts --emit ../../../weblate-localization/strings
node:internal/modules/esm/resolve:303
  return new ERR_PACKAGE_PATH_NOT_EXPORTED(
         ^
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './esm-register' is not defined by "exports" in /home/runner/work/specify7/specify7/specifyweb/frontend/js_src/node_modules/@swc-node/register/package.json imported from /home/runner/work/specify7/specify7/specifyweb/frontend/js_src/
    at exportsNotFound (node:internal/modules/esm/resolve:303:10)
    at packageExportsResolve (node:internal/modules/esm/resolve:650:9)
    at packageResolve (node:internal/modules/esm/resolve:836:14)
    at moduleResolve (node:internal/modules/esm/resolve:910:20)
    at defaultResolve (node:internal/modules/esm/resolve:1130:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:396:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:365:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:240:38)
    at ModuleLoader.import (node:internal/modules/esm/loader:328:34)
    at loadESM (node:internal/process/esm_loader:23:33) {
  code: 'ERR_PACKAGE_PATH_NOT_EXPORTED'
}
Node.js v20.11.0
Error: Process completed with exit code 1.
```

I still get a similar error using `esm` instead of `esm-register`
```
> node --import @swc-node/register/esm lib/localization/__tests__/scanUsages.ts --emit ../../../weblate-localization/strings
node:internal/modules/esm/get_format:160
  throw new ERR_UNKNOWN_FILE_EXTENSION(ext, filepath);
        ^
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /home/runner/work/specify7/specify7/specifyweb/frontend/js_src/lib/localization/__tests__/scanUsages.ts
    at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:160:9)
    at defaultGetFormat (node:internal/modules/esm/get_format:203:36)
    at defaultLoad (node:internal/modules/esm/load:143:22)
    at async ModuleLoader.load (node:internal/modules/esm/loader:409:7)
    at async ModuleLoader.moduleProvider (node:internal/modules/esm/loader:291:45)
    at async link (node:internal/modules/esm/module_job:76:21) {
  code: 'ERR_UNKNOWN_FILE_EXTENSION'
}
Node.js v20.11.0
Error: Process completed with exit code 1.
```

**Comment by acwhite211** [comment]
ok, so node got upgraded to >v20.6, so the directions are to use `node --import @swc-node/register/esm-register lib/localization/__tests__/scanUsages.ts` but it still seems to only work with using `node --loader @swc-node/register/esm lib/localization/__tests__/scanUsages.ts`

File: specifyweb/frontend/js_src/lib/components/Forms/DateRange.tsx [hunk 1]
```diff
@@ -9,22 +9,22 @@ import { parseAnyDate } from '../../utils/relativeDate';
...
 import { flippedSortTypes } from '../QueryBuilder/helpers';
-import { SpecifyResource } from '../DataModel/legacyTypes';
 import { runQuery } from '../QueryBuilder/ResultsWrapper';
 
 export function DateRange({
   table,
   filterQueryField,
 }: {
-  //   readonly table: SpecifyModel;
+  //   Readonly table: SpecifyModel;
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/Forms/DateRange.tsx:27]: not a fault of this PR, but could you please remove these commended out lines?
thanks

File: specifyweb/frontend/js_src/lib/components/Forms/DateRange.tsx [hunk 2]
```diff
@@ -44,7 +44,7 @@ function DateRangeComponent({
   dateFields,
   filterQueryField,
 }: {
-  //   readonly table: SpecifyModel;
+  //   Readonly table: SpecifyModel;
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/Forms/DateRange.tsx:47]: please remove this line

File: specifyweb/frontend/js_src/lib/components/Forms/DeleteBlocked.tsx [hunk 3]
```diff
@@ -152,7 +152,7 @@ function BlockerPreview({
     [ids]
   );
 
-  //   const table = parentRelationship?.relatedModel ?? directRelationship.model;
+  //   Const table = parentRelationship?.relatedModel ?? directRelationship.model;
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/Forms/DeleteBlocked.tsx:155]: please remove this line
thanks!

File: specifyweb/frontend/js_src/package.json [hunk 4]
```diff
@@ -12,14 +12,14 @@
...
-    "node": ">=18.0.0"
+    "node": ">=20.6.0"
   },
   "scripts": {
     "//": "All of these should be executed from the js_src directory",
     "localizationTests": "node --loader @swc-node/register/esm lib/localization/__tests__/scanUsages.ts",
     "pullFromWeblate": "node --loader @swc-node/register/esm lib/localization/utils/pull.ts",
     "schema:extract": "node --loader @swc-node/register/esm lib/localization/schema-localization/extract.ts",
-    "schema:pullFromWeblate": "node --loader @swc-node/register/esm lib/localization/schema-localization/pull.ts",
+    "schema:pullFromWeblate": "ts --loader @swc-node/register/esm lib/localization/schema-localization/pull.ts",
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/package.json:22]: hmm, was this supposed to be `node` rather than `ts`?

**Review by maxpatiiuk — COMMENTED** [review]
given how many commits this took, I am sorry you had to fight with it so much to get it to work
but thanks for updating us to newer node.js version

**Comment by maxpatiiuk** [comment]
README.md still mentions node.js 18 in two places:
https://github.com/specify/specify7/#:~:text=https%3A//deb.nodesource.com/setup_18.
https://github.com/specify/specify7/#:~:text=sure%20you%20have-,Node.js%2018,-installed%3A

could you please update?

**Comment by acwhite211** [comment]
This PR is looking good now.  Should be ready to merge into production.

File: README.md [hunk 5]
```diff
@@ -82,27 +82,36 @@ under GNU General Public License 2 (GPL2).
...
-            - [Turning on debugging](#turning-on-debugging)
-            - [The development server](#the-development-server)
-            - [The Specify 7 worker](#the-specify-7-worker)
-            - [Installing production requirements](#installing-production-requirements)
-            - [Setting up Apache](#setting-up-apache)
-            - [Restarting Apache](#restarting-apache)
-    - [Updating Specify 7](#updating-specify-7)
-    - [Updating the database (Specify 6) version](#updating-the-database-specify-6-version)
-    - [Localizing Specify 7](#localizing-specify-7)
+  - [Table of Contents](#table-of-contents)
```
Comment (maxpatiiuk) [inline README.md:85]: nice!
was this done by a vscode plugin?

**Review by maxpatiiuk — APPROVED** [review]
💎

**Review by melton-jason — APPROVED** [review]

**Review by CarolineDenis — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (acwhite211) [reply README.md:85]: ya, didn't notice that when I pushed the commit.

**Comment by maxpatiiuk** [comment]
Hmm, the [github action](https://github.com/specify/specify7/actions/runs/7963889013/job/21740345853) is still failing despite the changes in this PR:

Screenshot 2024-02-19 at 10 59 30

> ok, so node got upgraded to >v20.6, so the directions are to use node --import @swc-node/register/esm-register lib/localization/__tests__/scanUsages.ts but it still seems to only work with using node --loader @swc-node/register/esm lib/localization/__tests__/scanUsages.ts

That is because our local version of `@swc-node/register` is outdated.
The new feature is only available in the newer version - https://github.com/swc-project/swc-node/releases/tag/%40swc-node%2Fregister%401.8.0#:~:text=feat%3A%20add%20esm%2Dregister%20for%20node%3E20.6%20by%20%40yeliex%20in%20%23748

**Pull request merged**